Document iteration encodings (#1064)
openPMD/openPMD-api: docs/source/usage/concepts.rst

The collection of iterations is called a *Series*.
openPMD-api implements various file-formats (backends) and encoding strategies for openPMD Series, from simple one-file-per-iteration writes over using the backend-provided support for internal updates of records to data streaming techniques.

**Iteration encoding:** The openPMD-api can encode iterations in different ways.
The method ``Series::setIterationEncoding()`` (C++) or ``Series.set_iteration_encoding()`` (Python) may be used in writing for selecting one of the following encodings explicitly:

* **group-based iteration encoding:** This encoding is the default.
  It creates a separate group in the hierarchy of the openPMD standard for each iteration.
  As an example, all data pertaining to iteration 0 may be found in group ``/data/0``, for iteration 100 in ``/data/100``.
* **file-based iteration encoding:** A unique file on the filesystem is created for each iteration.

>   The preferred way to create a file-based iteration encoding is by specifying an expansion pattern in the constructor of the ``Series`` class.

Reviewer: ```suggestion
  The preferred way to create a file-based iteration encoding is by specifying an expansion pattern in the ``filepath`` argument of the constructor of the ``Series`` class.
```

>   Creating a ``Series`` by the filename ``series%T.json`` will create files ``series0.json``, ``series100.json`` and ``series200.json`` for iterations 0, 100 and 200.

Reviewer: I would recommend people to use `_` because that simplifies CLI handling of files.
Example: some people call their output series `diagnostic1`, `diagnostic2`, etc. and this can be confusing to parse back in reading for scripts and regex' that do file handling:
```suggestion
  Creating a ``Series`` by the filepath ``"series_%T.json"`` will create files ``series_0.json``, ``series_100.json`` and ``series_200.json`` for iterations 0, 100 and 200.
```

Author: The reason why I did not add the underscore is that it might mislead people into thinking that it's part of the syntax. On the other hand, probably no harm done if people believe that..

>   A padding may be specified by ``series%06T.json`` to create files ``series000000.json``, ``series000100.json`` and ``series000200.json``.

Reviewer: ```suggestion
  A padding may be specified by ``"series_%06T.json"`` to create files ``series_000000.json``, ``series_000100.json`` and ``series_000200.json``.
```

  The inner group layout of each file is identical to that of the group-based encoding.

> * **variable-based iteration encoding:** This encoding uses a feature of some backends (i.e. ADIOS2) to maintain datasets and attributes in several versions (i.e. to be *variable*).

Reviewer: ```suggestion
* **variable-based iteration encoding:** This experimental encoding uses a feature of some backends (i.e., ADIOS2) to maintain datasets and attributes in several versions (i.e., iterations are stored inside *variables*).
```

* **file-based iteration encoding:** A unique file on the filesystem is created for each iteration.
  The preferred way to create a file-based iteration encoding is by specifying an expansion pattern in the ``filepath`` argument of the constructor of the ``Series`` class.
  Creating a ``Series`` by the filepath ``"series_%T.json"`` will create files ``series_0.json``, ``series_100.json`` and ``series_200.json`` for iterations 0, 100 and 200.
  A padding may be specified by ``"series_%06T.json"`` to create files ``series_000000.json``, ``series_000100.json`` and ``series_000200.json``.
  The inner group layout of each file is identical to that of the group-based encoding.
* **variable-based iteration encoding:** This experimental encoding uses a feature of some backends (i.e., ADIOS2) to maintain datasets and attributes in several versions (i.e., iterations are stored inside *variables*).
  No iteration-specific groups are created and the corresponding layer is dropped from the openPMD hierarchy.
  In backends that do not support this feature, a series created with this encoding can only contain one iteration.

> Alternative spellings: groupbased, variablebased, filebased, group_based, variable_based, file_based

Reviewer: ```suggestion
Spellings for constants in the C++ (``IterationEncoding``) and Python (``Iteration_Encoding``) API: ``groupBased``, ``variableBased``, ``fileBased``, ``group_based``, ``group_based``, ``variable_based``
```
